src/type_parsers/datetime_parser.jl

```diff
@@ -49,108 +59,155 @@ function _unsafe_datetime(y=0, m=1, d=1, h=0, mi=0, s=0, ms=0)
 
-# [y]yyy-[m]m-[d]d((T|\s)HH:MM:SS(\.s{1,3}})?)?(zzzz|ZZZ|\Z)?
+# [-]y{1,10}-[m]m-[d]d((T|\s)HH:MM:SS(\.s{1,3})?)?(zzzz|ZZZ|\Z)?
 Base.@propagate_inbounds function _default_tryparse_timestamp(buf, pos, len, code, b, options)
     delim = options.delim.token
     cq = options.cq.token
     rounding = options.rounding
-    # ensure there is enough room for at least yyyy-m-d
-    if len - pos < 8
+    # ensure there is enough room for at least y-mm-dd
+    if len - pos + 1 < 7
         (b != delim) && (code |= Parsers.INVALID)
         (pos >= len) && (code |= Parsers.EOF)
         return _unsafe_datetime(0), code, pos
     end
+    sign_mul = 1
+    if b == UInt8('-')
+        sign_mul = -1
+        pos += 1
+        b = buf[pos]
+    end
 
     year = 0
-    for i in 1:4
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(0), code | Parsers.INVALID, pos)
-        year = Int(b) + 10 * year
-        b = buf[pos += 1]
-        (i > 2 && b == UInt8('-')) && break
+    for i in 1:10
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(0), code | Parsers.INVALID, pos)
+        year = Int(b0) + 10 * year
+        pos += 1
+        pos > len && (return _unsafe_datetime(0), code | Parsers.INVALID | Parsers.EOF, pos)
+        b = buf[pos]
+        b == UInt8('-') && break
     end
-    b != UInt8('-')  && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
-    b = buf[pos += 1]
+    year *= sign_mul
+    # If the year is larger than what can be represented by a 32 bit integer, fail to parse,
+    # values between typemin(Int32) and MIN_DATETIME are clamped to MIN_DATETIME
+    # values between typemax(Int32) and MAX_DATETIME are clamped to MAX_DATETIME
+    overflowed = (year > typemax(Int32) || year < typemin(Int32))
+    if b != UInt8('-') || overflowed
+        overflowed || (code |= Parsers.OVERFLOW)
+        return (_unsafe_datetime(year), code | Parsers.INVALID, pos)
+    end
+    pos += 1
+    pos > len && (return _unsafe_datetime(0), code | Parsers.INVALID | Parsers.EOF, pos)
+    b = buf[pos]
 
     month = 0
     for _ in 1:2
-        b -= 0x30
-        b > 0x09 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
-        month = Int(b) + 10 * month
-        b = buf[pos += 1]
-        b == UInt8('-') && break
+        b0 = b - 0x30
+        b0 > 0x09 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
+        month = Int(b0) + 10 * month
+        pos += 1
+        pos > len && (return _unsafe_datetime(year), code | Parsers.INVALID | Parsers.EOF, pos)
+        b = buf[pos]
     end
     month > 12 && (return _unsafe_datetime(year), code | Parsers.INVALID, pos)
-    b != UInt8('-')  && (return _unsafe_datetime(year, month), code | Parsers.INVALID, pos)
-    b = buf[pos += 1]
+    b != UInt8('-') && (return _unsafe_datetime(year, month), code | Parsers.INVALID, pos)
+    pos += 1
+    pos > len && (return _unsafe_datetime(year), code | Parsers.INVALID | Parsers.EOF, pos)
```

Comment:
why not also include the month here?
```suggestion
    pos > len && (return _unsafe_datetime(year, month), code | Parsers.INVALID | Parsers.EOF, pos)
```

Reply:
Yeah, we could 👍